### PR TITLE
[tests-only][full-ci] Added API tests for searching users by normal user using display name

### PR DESCRIPTION
**Pull request by ghost**
## Description

Adding test to get user data as a normal user using search queries
```cucumber
Scenario Outline: non-admin user searches other users by display name
Scenario Outline: non-admin user tries to search for a user with than 3 characters
Scenario Outline: non-admin user tries to search for a user by display name with invalid characters/token
```
and this PR also fixes some of the incorrect json assertion in file `tests/acceptance/features/apiGraph/getUser.feature`

## Related Issue

- Part of: https://github.com/owncloud/ocis/issues/7944
- https://github.com/owncloud/ocis/issues/8115

## How Has This Been Tested?

- test environment:
- test case 1:
- test case 2:
- ...

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Technical debt
- [ ] Tests only (no source changes)

## Checklist:

- [ ] Code changes
- [ ] Unit tests added
- [ ] Acceptance tests added
- [ ] Documentation ticket raised: <link>

### CODE REVIEW

File: tests/TestHelpers/GraphHelper.php [hunk 1]
```diff
@@ -259,6 +259,32 @@ public static function getUser(
...
+	 * @return ResponseInterface
+	 */
+	public static function searchUser(
+		string $baseUrl,
+		string $xRequestId,
+		string $adminUser,
+		string $adminPassword,
+		string $userName
+	): ResponseInterface {
+		$url = self::getFullUrl($baseUrl, 'users' . "?\$search=$userName");
```
Comment (saw-jan) [inline tests/TestHelpers/GraphHelper.php:278]: this needs to be a pattern. we can search users using other things like `displayName`, `mail`, `id`, or `groupTypes`
```suggestion
		$url = self::getFullUrl($baseUrl, 'users' . "?\$search=$searchTerm");
```

File: tests/acceptance/features/apiGraph/getUser.feature [hunk 2]
```diff
@@ -1327,3 +1327,104 @@ Feature: get users
       | Space Admin |
       | User        |
       | User Light  |
+
+
+  Scenario Outline: non-admin user tries to search for a user
```
Comment (saw-jan) [inline tests/acceptance/features/apiGraph/getUser.feature:1332]:
```suggestion
  Scenario Outline: non-admin user searches other user
```

File: tests/acceptance/features/apiGraph/getUser.feature [hunk 3]
```diff
@@ -1327,3 +1327,104 @@ Feature: get users
...
+    Given the administrator has assigned the role "<role>" to user "Alice" using the Graph API
+    And the administrator has assigned the role "<userRole>" to user "Brian" using the Graph API
+    When user "Brian" searches for user "<user>" using Graph API
```
Comment (saw-jan) [inline tests/acceptance/features/apiGraph/getUser.feature:1335]:
```suggestion
    When user "Brian" searches for user "ali" using Graph API
```

File: tests/acceptance/features/apiGraph/getUser.feature [hunk 4]
```diff
@@ -1327,3 +1327,104 @@ Feature: get users
...
+      | Space Admin | User Light  | ali  |
+      | Space Admin | Admin       | ali  |
+      | User        | Space Admin | ali  |
+      | User        | User        | ali  | 
+      | User        | User Light  | ali  |
+      | User        | Admin       | ali  |
+      | User Light  | Space Admin | ali  |
+      | User Light  | User        | ali  |
+      | User Light  | User Light  | ali  |
+      | User Light  | Admin       | ali  |
```
Comment (saw-jan) [inline tests/acceptance/features/apiGraph/getUser.feature:1388]:
```suggestion
      | userRole    | role        |
      | Space Admin | Space Admin |
      | Space Admin | User        |
      | Space Admin | User Light  |
      | Space Admin | Admin       |
      | User        | Space Admin |
      | User        | User        | 
      | User        | User Light  |
      | User        | Admin       |
      | User Light  | Space Admin |
      | User Light  | User        |
      | User Light  | User Light  |
      | User Light  | Admin       |
```

File: tests/acceptance/features/apiGraph/getUser.feature [hunk 5]
```diff
@@ -1327,3 +1327,104 @@ Feature: get users
...
+
+
+  Scenario Outline: non-admin user tries to search for a user with less characters
+    Given the administrator has assigned the role "<role>" to user "Alice" using the Graph API
+    And the administrator has assigned the role "<userRole>" to user "Brian" using the Graph API
+    When user "Brian" tries to search for user "<user>" using Graph API
```
Comment (SagarGi) [inline tests/acceptance/features/apiGraph/getUser.feature:1394]:
```suggestion
    When user "Brian" tries to search for user "ali" using Graph API
```

File: tests/acceptance/features/apiGraph/getUser.feature [hunk 6]
```diff
@@ -1327,3 +1327,104 @@ Feature: get users
...
+      | Space Admin | User Light  | al   |
+      | Space Admin | Admin       | al   |
+      | User        | Space Admin | al   |
+      | User        | User        | al   | 
+      | User        | User Light  | al   |
+      | User        | Admin       | al   |
+      | User Light  | Space Admin | al   |
+      | User Light  | User        | al   |
+      | User Light  | User Light  | al   |
+      | User Light  | Admin       | al   |
```
Comment (SagarGi) [inline tests/acceptance/features/apiGraph/getUser.feature:1430]:
```suggestion
      | userRole    | role        |
      | Space Admin | Space Admin |
      | Space Admin | User        |
      | Space Admin | User Light  |
      | Space Admin | Admin       |
      | User        | Space Admin |
      | User        | User        | 
      | User        | User Light  |
      | User        | Admin       |
      | User Light  | Space Admin |
      | User Light  | User        |
      | User Light  | User Light  |
      | User Light  | Admin       |
```
same here as well!

File: tests/TestHelpers/GraphHelper.php [hunk 7]
```diff
@@ -259,6 +259,32 @@ public static function getUser(
...
+	 * @return ResponseInterface
+	 */
+	public static function searchUser(
+		string $baseUrl,
+		string $xRequestId,
+		string $adminUser,
+		string $adminPassword,
+		string $searchTerm
+	): ResponseInterface {
+		$url = self::getFullUrl($baseUrl, 'users' . "?\$search=$searchTerm");
```
Comment (saw-jan) [inline tests/TestHelpers/GraphHelper.php:278]:
```suggestion
		$url = self::getFullUrl($baseUrl, "users?\$search=$searchTerm");
```

File: tests/acceptance/features/apiGraph/getUser.feature [hunk 8]
```diff
@@ -1327,3 +1327,104 @@ Feature: get users
...
+      | User        | User        |
+      | User        | User Light  |
+      | User        | Admin       |
+      | User Light  | Space Admin |
+      | User Light  | User        |
+      | User Light  | User Light  |
+      | User Light  | Admin       |
+
+
+  Scenario Outline: non-admin user tries to search for a user with less characters
```
Comment (saw-jan) [inline tests/acceptance/features/apiGraph/getUser.feature:1391]:
```suggestion
  Scenario Outline: non-admin user tries to search for a user with less than 3 characters
```

File: tests/acceptance/features/bootstrap/GraphContext.php [hunk 9]
```diff
@@ -1264,6 +1264,28 @@ public function userTriesToGetInformationOfUser(string $byUser, string $user): v
...
+	 * @When user :byUser tries to search for user :user using Graph API
+	 * @When user :byUser searches for user :user using Graph API
+	 *
+	 * @param string $byUser
+	 * @param string $user
+	 *
+	 * @return void
+	 * @throws GuzzleException
+	 */
+	public function userTriesToSearchForUser(string $byUser, string $user): void {
```
Comment (saw-jan) [inline tests/acceptance/features/bootstrap/GraphContext.php:1277]:
```suggestion
	 * @When user :byUser tries to search for user :searchTerm using Graph API
	 * @When user :byUser searches for user :searchTerm using Graph API
	 *
	 * @param string $byUser
	 * @param string $user
	 *
	 * @return void
	 * @throws GuzzleException
	 */
	public function userTriesToSearchForUser(string $byUser, string $searchTerm): void {
```

File: tests/acceptance/features/apiGraph/getUser.feature [hunk 10]
```diff
@@ -1327,3 +1344,163 @@ Feature: get users
...
+      | User        | User        |
+      | User        | User Light  |
+      | User        | Admin       |
+      | User Light  | Space Admin |
+      | User Light  | User        |
+      | User Light  | User Light  |
+      | User Light  | Admin       |
+  
+
+  Scenario Outline: non-admin user tries to search for a user by display name with invalid characters/token
```
Comment (saw-jan) [inline tests/acceptance/features/apiGraph/getUser.feature:1452]: Could you put issue tag if there's any bug report for such scenario?

File: tests/acceptance/features/apiGraph/getUser.feature [hunk 11]
```diff
@@ -75,8 +75,10 @@ Feature: get users
             "message"
           ],
           "properties": {
-            "type": "string",
-            "enum": ["Unauthorized"]
+            "message":{
+              "type": "string",
+              "enum": ["Unauthorized"]
+            }
```
Comment (saw-jan) [inline tests/acceptance/features/apiGraph/getUser.feature:81]: why was this change needed?

File: tests/acceptance/features/apiGraph/getUser.feature [hunk 12]
```diff
@@ -359,14 +363,21 @@ Feature: get users
...
+                    "required": [
+                      "displayName",
+                      "id"
+                    ],
+                    "properties": {
+                      "displayName": {
+                        "type": "string",
+                        "enum": [""]
+                      },
+                      "id": {
```
Comment (PrajwolAmatya) [inline tests/acceptance/features/apiGraph/getUser.feature:377]:
```suggestion
                    "required": [
                      "id"
                    ],
                    "properties": {
                      "id": {
```

The displayName is returned empty, so no need to assert it.

**Review by SagarGi — COMMENTED** [review]
others LGTM :+1:

File: tests/acceptance/features/apiGraph/getUser.feature (hunk 11 above)
Comment (saw-jan) [reply tests/acceptance/features/apiGraph/getUser.feature:81]: great! :+1:

**Comment by SwikritiT** [comment]
the line numbers in expected to fail needs to be changed other than that the changes looks good to me

File: tests/TestHelpers/GraphHelper.php (hunk 1 above)
Comment (saw-jan) [reply tests/TestHelpers/GraphHelper.php]: >we can search users using other things like `displayName`, `mail`, `id`, or `groupTypes`

I was mistaken here. this could or could not be true (I only know of displayname and mail)
The actual statement was that the search response should only contains these info

File: tests/acceptance/expected-failures-localAPI-on-OCIS-storage.md [hunk 13]
```diff
@@ -71,18 +69,20 @@ The expected failures in this file are from features in the owncloud/ocis repo.
...
 - [apiGraph/getUser.feature:653](https://github.com/owncloud/ocis/blob/master/tests/acceptance/features/apiGraph/getUser.feature#L653)
+- [apiGraph/getUser.feature:654](https://github.com/owncloud/ocis/blob/master/tests/acceptance/features/apiGraph/getUser.feature#L654)
+- [apiGraph/getUser.feature:655](https://github.com/owncloud/ocis/blob/master/tests/acceptance/features/apiGraph/getUser.feature#L655)
+- [apiGraph/getUser.feature:656](https://github.com/owncloud/ocis/blob/master/tests/acceptance/features/apiGraph/getUser.feature#L656)
+- [apiGraph/getUser.feature:657](https://github.com/owncloud/ocis/blob/master/tests/acceptance/features/apiGraph/getUser.feature#L657)
+- [apiGraph/getUser.feature:658](https://github.com/owncloud/ocis/blob/master/tests/acceptance/features/apiGraph/getUser.feature#L658)
+- [apiGraph/getUser.feature:659](https://github.com/owncloud/ocis/blob/master/tests/acceptance/features/apiGraph/getUser.feature#L659)
+- [apiGraph/getUser.feature:660](https://github.com/owncloud/ocis/blob/master/tests/acceptance/features/apiGraph/getUser.feature#L660)
```
Comment (saw-jan) [inline tests/acceptance/expected-failures-localAPI-on-OCIS-storage.md:84]: I believe these are just changes in line numbers, right?

**Review by saw-jan — APPROVED** [review]
LGTM

File: tests/acceptance/features/apiGraph/getUser.feature [hunk 14]
```diff
@@ -1327,3 +1339,163 @@ Feature: get users
       | Space Admin |
       | User        |
       | User Light  |
+
+
+  Scenario Outline: non-admin user searches other users by display name
+    Given the administrator has assigned the role "<role>" to user "Alice" using the Graph API
+    And the administrator has assigned the role "<userRole>" to user "Brian" using the Graph API
```
Comment (grgprarup) [inline tests/acceptance/features/apiGraph/getUser.feature:1346]:
```suggestion
    Given the administrator has assigned the role "<role>" to user "Alice" using the Graph API
```
What is the purpose of assigning roles to users? If searching has to be done with different roles then why assign roles to two users? Retext the scenario description.

File: tests/acceptance/features/apiGraph/getUser.feature [hunk 15]
```diff
@@ -1327,3 +1339,163 @@ Feature: get users
...
+              "type": "string",
+              "enum": ["Member"]
+            }
+          }
+        }
+      }
+    }
+    """
+    Examples:
+      | userRole    | role        |
```
Comment (grgprarup) [inline tests/acceptance/features/apiGraph/getUser.feature:1388]: use kebab case for example table header.

File: tests/acceptance/features/apiGraph/getUser.feature [hunk 16]
```diff
@@ -1327,3 +1339,129 @@ Feature: get users
...
+      | Alice-From-Wonderland | -From-Wonderland |
+      | Alice@From@Wonderland | @From@Wonderland |
+      | Alice-From-Wonderland | -From-Wonderland |
+      | Alice@From@Wonderland | @From@Wonderland |
+      | Alice-From-Wonderland | -From-Wonderland |
+      | Alice@From@Wonderland | @From@Wonderland |
+      | Alice-From-Wonderland | -From-Wonderland |
+      | Alice@From@Wonderland | @From@Wonderland |
+      | Alice-From-Wonderland | -From-Wonderland |
+      | Alice@From@Wonderland | @From@Wonderland |
```
Comment (saw-jan) [inline tests/acceptance/features/apiGraph/getUser.feature:1467]: I think we only need two of them
```suggestion
      | Alice-From-Wonderland | -From-Wonderland |
      | Alice@From@Wonderland | @From@Wonderland |
```

**Review by saw-jan — CHANGES_REQUESTED** [review]

**Review by saw-jan — DISMISSED** [review]

**Review by SagarGi — APPROVED** [review]
LGTM :+1:

**Comment by saw-jan** [comment]
Please, rebase the PR

**Review by saw-jan — APPROVED** [review]
